### PR TITLE
Sine Skewed distribution

### PR DESCRIPTION
**Pull request by OlaRonning**
Translating Sine Skewed distribution from [Pyro](https://github.com/pyro-ppl/pyro/pull/2826).

### CODE REVIEW

File: numpyro/distributions/directional.py [hunk 1]
```diff
@@ -19,6 +19,143 @@
...
+    So for the above snippet it must hold that::
+
+        skew_phi.abs()+skew_psi.abs() <= 1
+
+    We handle this in the prior by computing psi_bound and use it to scale skew_psi.
+    We do **not** use psi_bound as::
+
+        skew_psi = pyro.sample('skew_psi', Uniform(-psi_bound, psi_bound))
+
+    as it would make the support for the Uniform distribution dynamic.
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:69]: I'll add a PR for l1_ball to resolve this issue. In my opinion, this is quite inconvenient for users.

File: numpyro/distributions/directional.py [hunk 2]
```diff
@@ -19,6 +19,143 @@
...
+        :class:`~numnumpyro.distributions.SineBivariateVonMises`, 1D :class:`~numpyro.distributions.ProjectedNormal`,
+        and :class:`~numpyro.distributions.Uniform` (-pi, pi).
+    :param torch.tensor skewness: skewness of the distribution.
+    """
+
+    arg_constraints = {
+        "skewness": constraints.independent(constraints.interval(-1.0, 1.0), 1)
+    }
+
+    support = constraints.independent(constraints.real, 1)
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:94]: similar to the other PR, add a TODO for further discussion

File: numpyro/distributions/directional.py [hunk 3]
```diff
@@ -19,6 +19,143 @@
...
+
+    def __init__(self, base_dist: Distribution, skewness, validate_args=None):
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:96]: assert base_dist.event_shape == ...

**Review by fehiepsi — COMMENTED** [review]
LTGM overall. I'll add a PR for `l1_norm` constraint shortly.

File: numpyro/distributions/directional.py (hunk 1 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:69]: Thanks, that would be nice!

File: numpyro/distributions/directional.py (hunk 3 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:96]: Make sense.

File: numpyro/distributions/directional.py (hunk 2 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:94]: done C:

File: numpyro/distributions/directional.py [hunk 4]
```diff
@@ -19,6 +19,149 @@
...
+        return samples
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        if self.base_dist._validate_args:
+            self.base_dist._validate_sample(value)
+
+        # Eq. 2.1 in [1]
+        skew_prob = jnp.log(
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:152]: why not use `log1p` as in Pyro?

File: numpyro/distributions/directional.py [hunk 5]
```diff
@@ -19,6 +19,149 @@
...
+
+    ** References: **
+        1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+            Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param torch.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
+        distributions include: 1D :class:`~numpyro.distributions.VonMises`,
+        :class:`~numnumpyro.distributions.SineBivariateVonMises`, 1D :class:`~numpyro.distributions.ProjectedNormal`,
+        and :class:`~numpyro.distributions.Uniform` (-pi, pi).
+    :param torch.tensor skewness: skewness of the distribution.
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:87]: typo `torch.tensor` -> `np.ndarray`

File: numpyro/distributions/directional.py [hunk 6]
```diff
@@ -19,6 +19,149 @@
...
+    .. note:: An event in the base distribution must be on a d-torus, so the event_shape must be (d,).
+
+    .. note:: For the skewness parameter, it must hold that the sum of the absolute value of its weights for an event
+        must be less than or equal to one. See eq. 2.1 in [1].
+
+    ** References: **
+        1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+            Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param torch.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:83]: typo: torch

File: numpyro/distributions/directional.py [hunk 7]
```diff
@@ -19,6 +19,149 @@
...
+    base distribution.
+
+    Torus distributions are distributions with support on products of circles
+    (i.e., :math:`\bigotimes^d S^1` where :math:`S^1=[-\pi, \pi)` ). So, a 0-torus is a point, the 1-torus is a circle,
+    and the 2-torus is commonly associated with the donut shape.
+
+    The Sine Skewed X distribution is parameterized by a weight parameter for each dimension of the event of X.
+    For example with a von Mises distribution over a circle (1-torus), the Sine Skewed von Mises Distribution has one
+    skew parameter. The skewness parameters can be inferred using :class:`~numpyro.infer.HMC` or
+    :class:`~numpyro.infer.NUTS`. For example, the following will produce a uniform prior over
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:34]: `numpyro.infer.mcmc.NUTS`

File: numpyro/distributions/directional.py [hunk 8]
```diff
@@ -19,6 +19,149 @@
...
+
+        skew_phi.abs()+skew_psi.abs() <= 1
+
+    We handle this in the prior by computing psi_bound and use it to scale skew_psi.
+    We do **not** use psi_bound as::
+
+        skew_psi = pyro.sample('skew_psi', Uniform(-psi_bound, psi_bound))
+
+    as it would make the support for the Uniform distribution dynamic.
+    In the context of :class:`~pyro.infer.SVI`, this distribution can freely be used as a likelihood. But, when used as
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:70]: `numpyro.infer.svi.SVI`

File: numpyro/distributions/directional.py [hunk 9]
```diff
@@ -19,6 +19,149 @@
...
+    distribution the skewness parameters are constraint. The constraint requires the sum of the absolute values of
+    skewness to be less than or equal to one.
+    So for the above snippet it must hold that::
+
+        skew_phi.abs()+skew_psi.abs() <= 1
+
+    We handle this in the prior by computing psi_bound and use it to scale skew_psi.
+    We do **not** use psi_bound as::
+
+        skew_psi = pyro.sample('skew_psi', Uniform(-psi_bound, psi_bound))
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:67]: `numpyro`

File: numpyro/distributions/directional.py [hunk 10]
```diff
@@ -19,6 +19,149 @@
...
+            psi_loc = numpyro.sample('psi_loc', VonMises(-pi / 2, 2.))
+            phi_conc = numpyro.sample('phi_conc', Beta(2., 2.))
+            psi_conc = numpyro.sample('psi_conc', Beta(2., 2.))
+            corr_scale = numpyro.sample('corr_scale', Beta(2., 5.))
+
+            # SS prior
+            skew_phi = numpyro.sample('skew_phi', Uniform(-1., 1.))
+            psi_bound = 1 - skew_phi.abs()
+            skew_psi = numpyro.sample('skew_psi', Uniform(-1., 1.))
+            skewness = torch.stack((skew_phi, psi_bound * skew_psi), dim=-1)
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:49]: torch.stack

File: numpyro/distributions/directional.py [hunk 11]
```diff
@@ -19,6 +19,149 @@
...
+                                             psi_concentration=1000 * psi_conc,
+                                             weighted_correlation=corr_scale)
+                return numpyro.sample('phi_psi', SineSkewed(sine, skewness), obs=obs)
+
+    To ensure the skewing does not alter the normalization constant of the (Sine Bivaraite von Mises) base
+    distribution the skewness parameters are constraint. The constraint requires the sum of the absolute values of
+    skewness to be less than or equal to one.
+    So for the above snippet it must hold that::
+
+        skew_phi.abs()+skew_psi.abs() <= 1
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:62]: abs(skew_phi) + abs(sknew_psi) <= 1

numpy ndarray does not have `abs` method

File: numpyro/distributions/directional.py [hunk 12]
```diff
@@ -19,6 +19,149 @@
...
+        must be less than or equal to one. See eq. 2.1 in [1].
+
+    ** References: **
+        1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+            Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param torch.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
+        distributions include: 1D :class:`~numpyro.distributions.VonMises`,
+        :class:`~numnumpyro.distributions.SineBivariateVonMises`, 1D :class:`~numpyro.distributions.ProjectedNormal`,
+        and :class:`~numpyro.distributions.Uniform` (-pi, pi).
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:86]: Use full path for those distributions

File: numpyro/distributions/directional.py (hunk 4 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:152]: I copied over a commit without all revisions. Sorry for making you review the same thing twice.

File: numpyro/distributions/directional.py (hunk 5 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:87]: thanks, I'll fix this.

File: numpyro/distributions/directional.py (hunk 7 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:34]: I'll update this.

File: numpyro/distributions/directional.py (hunk 12 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py]: done

File: numpyro/distributions/directional.py (hunk 8 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py]: fixed.

File: numpyro/distributions/directional.py (hunk 10 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py]: thanks, fixed it.

File: numpyro/distributions/directional.py [hunk 13]
```diff
@@ -153,6 +153,138 @@ def variance(self):
...
+
+    :param numpyro.distributions.Distribution base_dist: base density on a d-dimensional torus. Supported base
+        distributions include: 1D :class:`~numpyro.distributions.VonMises`,
+        :class:`~numnumpyro.distributions.SineBivariateVonMises`, 1D :class:`~numpyro.distributions.ProjectedNormal`,
+        and :class:`~numpyro.distributions.Uniform` (-pi, pi).
+    :param jax.numpy.array skewness: skewness of the distribution.
+    """
+
+    arg_constraints = {
+        "skewness": constraints.independent(constraints.interval(-1.0, 1.0), 1)
```
Comment (fehiepsi) [inline numpyro/distributions/directional.py:216]: I think we can use `constraints.l1_ball` here.

**Review by fehiepsi — DISMISSED** [review]
Looks great to me! Thanks for porting this over NumPyro, Ola!

File: numpyro/distributions/directional.py (hunk 13 above)
Comment (OlaRonning) [reply numpyro/distributions/directional.py:216]: That makes sense; I'll update it.

**Comment by OlaRonning** [comment]
Thanks for all the reviews and feedback! It's a great help. 

I added `gen_values_outside_bounds` to be from either `[-2,-1.1]` or  `[1.1, 2.] `, but never both. I think this is fine. `gen_values_inside_bounds` generates from `[0., .5]` or `[-0.5, 0]`. This will need to be changed if we use the `l1_ball` constraint for event_dim > 2 in our testing. Let me know if I should make it more general.

**Review by fehiepsi — APPROVED** [review]
Yup, those testing values look reasonable to me. Thanks for explaining!

**Pull request merged**